File: pkg/objc/object.go
// Copyright (c) 2012 The 'objc' Package Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Package objc implements access to the Objective-C runtime from Go
package objc

import (
	"C"
	"unsafe"
)

// An Object represents an Objective-C object, along with
// some convenience methods only found on NSObjects.
type Object interface {
	// SendMsg sends an arbitrary message to the method on the
	// object that is idenfieid by selectorName.
	SendMsg(selectorName string, args ...interface{}) Object

	// SendSuperMsg is like SendMsg, but sends to the object's
	// super class instead.
	SendSuperMsg(selectorName string, args ...interface{}) Object

	// Alloc sends the  "alloc" message to the object.
	Alloc() Object

	// Init sends the "init" message to the object.
	Init() Object

	// Retain sends the "retain" message to the object.
	Retain() Object

	// Release sends the "release" message to the object.
	Release() Object

	// AutoRelease sends the "autorelease" message to the object.
	AutoRelease() Object

	// Copy sends the "copy" message to the object.
	Copy() Object

	// String returns a string-representation of the object.
	// This is equivalent to sending the "description"
	// message to the object, except that this method
	// returns a Go string.
	String() string

	// Pointer returns the in-memory address of the object.
	Pointer() uintptr

	// Uint returns the value of the object as an uint64.
	Uint() uint64

	// Int returns the value of the object as an int64.
	Int() int64

	// Float returns the value of the object as a float64.
	Float() float64

	// Bool returns the value of the object as a bool.
	Bool() bool
}

// Type object is the package's internal representation of an Object.
// Besides implementing the Objct interface, object also implements
// the Class interface.
type object struct {
	ptr uintptr
}

// Return the Object as a uintptr.
//
// Using package unsafe, this uintptr can further
// be converted to an unsafe.Pointer.
func (obj object) Pointer() uintptr {
	return obj.ptr
}

func (obj object) Alloc() Object {
	return obj.SendMsg("alloc")
}

func (obj object) Init() Object {
	return obj.SendMsg("init")
}

func (obj object) Retain() Object {
	return obj.SendMsg("retain")
}

func (obj object) Release() Object {
	return obj.SendMsg("release")
}

func (obj object) AutoRelease() Object {
	return obj.SendMsg("autorelease")
}

func (obj object) Copy() Object {
	return obj.SendMsg("copy")
}

func (obj object) String() string {
	pool := GetClass("NSAutoreleasePool").Alloc().Init()
	defer pool.Release()

	descString := obj.SendMsg("description")
	utf8Bytes := descString.SendMsg("UTF8String")
	if utf8Bytes.Pointer() != 0 {
		return C.GoString((*C.char)(unsafe.Pointer(utf8Bytes.Pointer())))
	}

	return "(nil)"
}

File: objc/object.m
// Copyright 2021 Liu Dong. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

#import <stdlib.h>
#import <stdint.h>
#import <stdbool.h>
#import <objc/NSObject.h>
#include "_cgo_export.h"
#import <objc/runtime.h>

void* C_NSObject_NewObject() {
    NSObject* result_ = [NSObject new];
    return result_;
}

bool Object_IsKindOfClass(void* ptr, void* classPtr) {
    NSObject* obj = (NSObject*)ptr;
    return [obj isKindOfClass:(Class)classPtr];
}

bool Object_IsMemberOfClass(void* ptr, void* classPtr) {
    NSObject* obj = (NSObject*)ptr;
    return [obj isMemberOfClass:(Class)classPtr];
}

bool Object_RespondsToSelector(void* ptr, void* sel) {
    NSObject* obj = (NSObject*)ptr;
    return [obj respondsToSelector:(SEL)sel];
}

bool Object_ConformsToProtocol(void* ptr, void* protocolPtr) {
    NSObject* obj = (NSObject*)ptr;
    return [obj conformsToProtocol:(Protocol*)protocolPtr];
}

void* Object_GetClass(void* ptr) {
    return object_getClass((id)ptr);
}

bool Object_IsProxy(void* ptr) {
    NSObject* obj = (NSObject*)ptr;
    return [obj isProxy];
}

int Object_RetainCount(void* ptr) {
    NSObject* obj = (NSObject*)ptr;
    return [obj retainCount];
}

void* Object_Retain(void* ptr) {
    NSObject* obj = (NSObject*)ptr;
    return [obj retain];
}

void Object_Release(void* ptr) {
    NSObject* obj = (NSObject*)ptr;
    [obj release];
}

void* Object_Autorelease(void* ptr) {
    NSObject* obj = (NSObject*)ptr;
    return [obj autorelease];
}

void* C_NSObject_Copy(void* ptr) {
    NSObject* nSObject = (NSObject*)ptr;
    id result_ = [nSObject copy];
    return result_;
}

void* C_NSObject_MutableCopy(void* ptr) {
    NSObject* nSObject = (NSObject*)ptr;
    id result_ = [nSObject mutableCopy];
    return result_;
}

void Object_Dealloc(void* ptr) {
    NSObject* obj = (NSObject*)ptr;
    [obj dealloc];
}

void* Object_PerformSelector(void* ptr, void* sel_p) {
    NSObject* obj = (NSObject*)ptr;
    return [obj performSelector:(SEL)sel_p];
}

void* Object_PerformSelector_WithObject(void* ptr, void* sel_p, void* param) {
    NSObject* obj = (NSObject*)ptr;
    return [obj performSelector:(SEL)sel_p withObject:(NSObject*)param];
}

void* Object_PerformSelector_WithObject_WithObject(void* ptr, void* sel_p, void* param1, void* param2) {
    NSObject* obj = (NSObject*)ptr;
    return [obj performSelector:(SEL)sel_p withObject:(NSObject*)param1  withObject:(NSObject*)param2];
}

const char* Object_Description(void* ptr) {
    return (const char*)[[((NSObject*)ptr) description] performSelector:@selector(UTF8String)];
}